controlplane/kubeadm/controllers/controller.go

```diff
@@ -254,6 +255,15 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		return ctrl.Result{}, nil
 	}
 
+	// All infrastructure machines created by KCP have infra-template ref annotation, but the ones that are adopted or created before this change do not.
+	// Owned machines with infra ref that doesn't have infra-template ref annotations are being patched here.
+	// This needs to be called after adoptMachines and before MachinesNeedingUpgrade() to avoid unnecessary upgrade
+	// on the newly adopted machines and older version machines.
+	if err := r.ReconcileMachineRefs(ctx, kcp.Spec.InfrastructureTemplate, ownedMachines); err != nil {
+		logger.Info("Failed to set infrastructure-reference annotation to the owned machines' infrastructure object")
+		return ctrl.Result{}, nil
```

Comment:
I think it's important to return here if we set any annotations – as I recall, the resource cache isn't "write through," so we won't see the annotation we're adding until after we get a watch notification (which updates the cache).

Reply:
Good catch!